enhancements/network/in-cluster-dns-and-loadbalancers-on-more-platforms.md

```diff
@@ -0,0 +1,320 @@
+### Implementation Details/Notes/Constraints
+
+The `InstallConfig` will gain new settings for `InClusterLoadBalancer` and
+`InternalDNS`. They are shown below, added to the existing settings for the
+External platform type.
+
+```
+type InClusterLoadBalancer struct {
+    // APIVIPs contains the VIP(s) to use for internal API communication. In
+    // dual stack clusters it contains an IPv4 and IPv6 address, otherwise only
+    // one VIP
+    //
+    // +kubebuilder:validation:MaxItems=2
+    // +kubebuilder:validation:UniqueItems=true
+    // +kubebuilder:validation:Format=ip
+    APIVIPs []string `json:"apiVIPs,omitempty"`
+
+    // IngressVIPs contains the VIP(s) to use for ingress traffic. In dual stack
+    // clusters it contains an IPv4 and IPv6 address, otherwise only one VIP
+    //  
+    // +kubebuilder:validation:MaxItems=2
+    // +kubebuilder:validation:UniqueItems=true
+    // +kubebuilder:validation:Format=ip
+    IngressVIPs []string `json:"ingressVIPs,omitempty"`
+}
+
+
+// Platform stores configuration related to external cloud providers.
+type Platform struct {
+    // PlatformName holds the arbitrary string representing the infrastructure
+    // provider name, expected to be set at the installation time. This field
+    // is solely for informational and reporting purposes and is not expected
+    // to be used for decision-making.
+    // +kubebuilder:default:="Unknown"
+    // +default="Unknown"
+    // +kubebuilder:validation:XValidation:rule="oldSelf == 'Unknown' || self == oldSelf",message="platform name cannot be changed once set"
+    // +optional
+    PlatformName string `json:"platformName,omitempty"`
+
+    // CloudControllerManager when set to external, this property will enable
+    // an external cloud provider.
+    // +kubebuilder:default:=""
+    // +default=""
+    // +kubebuilder:validation:Enum="";External
+    // +optional
+    CloudControllerManager CloudControllerManager `json:"cloudControllerManager,omitempty"`
+
+    // InClusterLoadBalancer is an optional feature that uses haproxy and
+    // keepalived as loadbalancers running in the cluster. Is is useful in
+    // environments where it is not possible or desirable to use loadbalancers outside
+    // of the cluster.
+    // +optional
+    InClusterLoadBalancer *InClusterLoadBalancer `json:"inClusterLoadBalancer,omitempty"`
+
+    // InternalDNS, when set, activates a DNS service running inside the cluster
+    // to provide DNS resolution internally. It is useful in environments where
+    // it is not possible or desirable to manage the cluster's internal DNS
+    // records in an external DNS system.
+    // +kubebuilder:default:=""
+    // +default=""
+    // +kubebuilder:validation:Enum="";CoreDNS
+    // +optional
+    InternalDNS InternalDNS `json:"internalDNS,omitempty"`
+}
+
+type InternalDNSType string
+
+const (
+    // CoreDNS is the default service used to implement internal DNS within a cluster.
+    CoreDNS InternalDNS = "CoreDNS"
```

Comment:
If we go with the loadBalancer field discussed above, we probably want this to have a default value of "None" or "UserManaged" or something like that to match the loadBalancer behavior.

We also may not want to use CoreDNS in the option name. There have been discussions about changing the implementation of the on-prem DNS to something lighter weight and it would be confusing if the option CoreDNS actually deployed dnsmasq. That's also why the loadBalancer parameter names are "OpenShiftManaged" and "UserManaged" instead of "keepalived" and "none".